semantic-scope-graph/src/Control/Carrier/Sketch/Fresh.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | This carrier interprets the Sketch effect, keeping track of
+-- the current scope and in-progress graph internally.
+module Control.Carrier.Sketch.Fresh
+  ( SketchC (..)
+  , runSketch
+  , module Control.Effect.Sketch
+  ) where
+
+import           Control.Algebra
+import           Control.Carrier.Fresh.Strict
+import           Control.Carrier.State.Strict
+import           Control.Effect.Sketch
+import           Control.Monad.IO.Class
+import           Data.Bifunctor
+import           Data.Module
+import           Data.Name (Name)
+import qualified Data.Name
+import           Data.ScopeGraph (ScopeGraph)
+import qualified Data.ScopeGraph as ScopeGraph
+import           Data.Semilattice.Lower
+import           Source.Span
+import qualified System.Path as Path
+
+data Sketchbook address = Sketchbook
+  { sGraph        :: ScopeGraph address
+  , sCurrentScope :: address
+  } deriving (Eq, Show)
+
+instance Lower (Sketchbook Name) where
+  lowerBound =
+    let
+      initialGraph = ScopeGraph.insertScope n initialScope lowerBound
+      initialScope = ScopeGraph.Scope mempty mempty mempty
+      n = Data.Name.nameI 0
+    in
+      Sketchbook initialGraph n
+
+newtype SketchC address m a = SketchC (StateC (Sketchbook address) (FreshC m) a)
+  deriving (Applicative, Functor, Monad, MonadIO)
+
+instance (Effect sig, Algebra sig m) => Algebra (Sketch Name :+: sig) (SketchC Name m) where
```

Comment:
If we’re neither using nor vending the `Fresh` effect here, and we’re therefore not e.g. using `gensym` any more, does it follow that we can remove `FreshC` from the carrier?

Reply:
Josh’s upcoming work uses gensym heavily, so I’d like to keep this here for now.

Reply:
Cool, that works for me 👍